src/main/java/app/linguistai/bmvp/service/gamification/quest/QuestService.java

```diff
@@ -0,0 +1,334 @@
+
+            // Step 4. Assign a "Send Message" quest
+            Integer randomSendMessagedTimes = this.generateRandomTimesForUseWordQuest();
+            Quest sendMessage = this.buildSendMessageQuest(user, randomSendMessagedTimes);
+
+            // Step 5. Save all quests to the repository
+            List<Quest> questsToAssign = new ArrayList<>(useWordQuests);
+            questsToAssign.add(sendMessage);
+
+            questRepository.saveAll(questsToAssign);
+        }
+        catch (NotFoundException e) {
+            log.error("User is not found for email {}", email);
+            throw e;
+        }
+        catch (Exception e) {
+            log.error("Assign quests failed for email {}", email, e);
+            throw new SomethingWentWrongException();
+        }
+    }
+
+    private List<Quest> buildBulkUseWordQuests(User user, Integer count) throws Exception {
+        if (count <= 0) {
+            throw new IllegalArgumentException("ERROR: Build bulk use word quests method failed. Count must be greater than zero.");
+        }
+
+        List<Quest> useWordQuests = new ArrayList<>();
+        HashSet<String> usedWords = new HashSet<>();
+
+        for (int i = 0; i < count; i++) {
+            // Get one active unknown word list
+            UnknownWordList randomActiveList = unknownWordService.getRandomActiveUnknownWordList(user.getId());
+
+            // Select a random word from the retrieved unknown word list, while ensuring no duplicates exist for quests
+            String randomWord = this.returnUniqueRandomWordForUseWordQuest(randomActiveList.getListId(), usedWords);
+            usedWords.add(randomWord);
+
+            // Generate a random "times" field
+            Integer randomUseWordTimes = this.generateRandomTimesForUseWordQuest();
+
+            // Build the Use Word quest and add to the results list
+            useWordQuests.add(this.buildUseWordQuest(user, randomWord, randomUseWordTimes));
+        }
+
+        return useWordQuests;
+    }
+
+    /**
+     * Select a random word from the retrieved unknown word list, while ensuring no duplicates exist for quests
+     * @param listId List ID for which the words will be pulled from
+     * @param usedWordsSoFar Used words so far in the generation
+     * @return Random unique word for "Use Word" type quests
+     * @throws Exception When a random word cannot be retrieved from list with ID listId
+     */
+    private String returnUniqueRandomWordForUseWordQuest(UUID listId, HashSet<String> usedWordsSoFar) throws Exception {
+        final int HARDCODED_DO_WHILE_LIMIT = 100; // to prevent infinite loops
+        int currentIteration = 0;
+        String randomWord;
+
+        do {
+            randomWord = unknownWordService.getRandomWordFromList(listId);
+            currentIteration++;
+        } while (usedWordsSoFar.contains(randomWord) && currentIteration < HARDCODED_DO_WHILE_LIMIT);
```

Comment:
methinks you can use the `List<String> getRandomWordFromList(UUID listId, Integer numOfWords) ` method from the UnknownWordService - this hardcoded do while limit business seems unnecessary

Reply:
In that case, checking for duplicate words become more costly. That's why I get the words one by one